src/coupled_driver.cpp

```diff
@@ -267,216 +283,277 @@ bool CoupledDriver::is_converged()
-
+  // On heat rank, update the elements' heat sources based on the cell-avged heat sources
   if (heat.active()) {
-    // Determine displacement for this rank
-    auto displacement = heat.local_displs_.at(heat.comm_.rank);
-    int n_local_elem = heat.n_local_elem();
-    // Set heat source in every element
-    for (int32_t local_elem = 0; local_elem < n_local_elem; ++local_elem) {
-      int32_t global_elem = local_elem + displacement;
-      // Get heat source for this element
-      CellHandle cell = elem_to_cell_.at(global_elem);
-      err_chk(heat.set_heat_source_at(local_elem, heat_source_.at(cell)),
-              "Error setting heat source for local element " +
-                std::to_string(local_elem));
+    if (relax) {
+      if (alpha_ == ROBBINS_MONRO) {
+        int n = i_picard_ + 1;
+        cell_heat_ = cell_heat_ / n + (1. - 1. / n) * cell_heat_prev_;
+      } else {
+        cell_heat_ = alpha_ * cell_heat_ + (1.0 - alpha_) * cell_heat_prev_;
+      }
+    }
+    for (gsl::index i = 0; i < cells_.size(); ++i) {
+      for (const auto& e : cell_to_elems_.at(cells_.at(i))) {
+        heat.set_heat_source_at(e, cell_heat_.at(i));
+      }
     }
   }
 }
 
 void CoupledDriver::update_temperature(bool relax)
 {
   comm_.message("Updating temperature");
-
   auto& neutronics = this->get_neutronics_driver();
   auto& heat = this->get_heat_driver();
 
-  // *************************************************************************
-  // Gather temperature on heat root and apply underrelaxation
-  // *************************************************************************
-
-  if (relax && comm_.rank == heat_root_) {
-    // Store previous temperature solution; a previous solution will always be present
-    // because a temperature IC is set and the neutronics solver runs first
-    std::copy(temperatures_.begin(), temperatures_.end(), temperatures_prev_.begin());
+  // Step 1: On each heat rank, assign the current iterate of local cell-avged T
+  // to the previous iterate
+  if (relax && heat.active()) {
+    std::copy(cell_temperatures_.begin(),
+              cell_temperatures_.end(),
+              cell_temperatures_prev_.begin());
   }
 
-  temperatures_ = heat.temperature();
-
-  if (relax && comm_.rank == heat_root_) {
-    if (alpha_T_ == ROBBINS_MONRO) {
-      int n = i_picard_ + 1;
-      temperatures_ = temperatures_ / n + (1. - 1. / n) * temperatures_prev_;
-    } else {
-      temperatures_ = alpha_T_ * temperatures_ + (1.0 - alpha_T_) * temperatures_prev_;
+  // Step 2: On each heat, compute cell-avged T
+  if (heat.active()) {
+    auto elem_temperatures = heat.temperature_local();
+    for (gsl::index i = 0; i < cells_.size(); ++i) {
+      double T_avg = 0.0;
+      double V_tot = 0.0;
+      for (const auto& e : cell_to_elems_.at(cells_.at(i))) {
+        double T = elem_temperatures.at(e);
+        double V = elem_volumes_.at(e);
+        T_avg += T * V;
+        V_tot += V;
+      }
+      T_avg /= V_tot;
```

Comment:
I'm glad you mentioned this, because it's something I had considered at the very beginning of this implementation.  Ideally, the cell volume from the neutronics model (`cell_volumes_[i]`) is identical to the sum of the element volumes from the heat/fluids model (`V_tot`).  However, this won't be the case if the models are slightly non-conformal or the assigned volumes in the neutronics model are a bit off.  

So, for the purposes of computing the volume-averaged temperature, I thought it was more self-consistent to use both the temperatures and volumes from the heat/fluids model.

Reply:
On the other hand, I'd grant non-conformity and volume inaccuracies are situations we shouldn't account for.  They cause lots of other problems and hence should be addressed for many other reasons.  And now that I'm re-reading the code, it is pretty convoluted to use V_tot.  I'll just use `cell_volumes_[i]` as you suggested.

Reply:
Hey, I understand what you meant here, and I see that `V_tot` and `cell_volumes_[i]` would identical.  I forgot the semantics of `cell_volumes_` until I looked back at the code:

```
void CoupledDriver::init_volumes()
{
  comm_.message("Initializing volumes");
  const auto& heat = this->get_heat_driver();
  const auto& neutronics = this->get_neutronics_driver();

  if (heat.active()) {
    elem_volumes_ = heat.volume_local();
    for (const auto& c : cells_) {
      double V = 0.0;
      for (const auto& e : cell_to_elems_.at(c)) {
        V += elem_volumes_.at(e);
      }
      cell_volumes_.push_back(V);
    }
  }
}
```

Reply:
Fixed in https://github.com/enrico-dev/enrico/pull/131/commits/faecec3e5c3a4c81222f0da12657df095342d84e